### DISCUSSION TITLE
UX improvements for Relationship field

### DISCUSSION THREAD

**Discussion by intoeetive** [ExpressionEngine/ExpressionEngine#2889] [Feature Requests]
Please use this issue to add (and discuss) the ideas on how the UX of Relationship field can be improved

**Comment by aasun** [comment]
I'm working on an extremely large and complex information/research site.  There are hundreds of thousands of entries with multiple relationship fields.  When more than a few relationship fields exist in a single entry, performance in the CP is greatly impacted.  So, some thoughts on how I'd like relationship fields to be able to be configured....

With tens of thousands of entries that can be related, it's not practical to set the view limit to all entries (it takes too long for the the entry edit screen to load).  Setting the limit to a small number (5 or 10) provides dramatic improvements for the entry edit page to load.  And the AJAX search in the relationship field will find matching entries from the tens of thousands, not just what is initially visible.  Great!  BUT, the  number of matching results that are displayed are limited by the display limit of the dropdown. Not great.  With tens of thousands of entries, there may be hundreds of entries that match a search.  As you continue to type your search term/phrase, the number of matches reduce. Great!  But, as a content editor, you won't know if your seeing all of the possible search matches until you see fewer results than what the view limit is on the dropdown.  Not great!

So, our current options are to increase the view limit of the relationship dropdowns to display more than what we think the search results may show.. In a site with tens of thousands of entries per relationship field, that takes us to showing hundreds in the dropdown. Which, completely bogs down the load time of the entry edit form. (with 10 or more relationship fields, it's nearly unusable).  Or, set the relationship field display limit to 5 or 10 and tell the content editors to open up the channel listing of the relationships to "pre-search" so they know exactly the title of the entry they want to relate so they have a very long search term they can type in the relationship field to find the entry to relate.  Which, is a completely impractical workflow.

So, Ideally, I'd like to be able to set the relationship drop down (initial) display limit to 0. (there is no practical reason to show any, when the relationship field pulls from thousands of entries..  Browsing a dropdown will never be the workflow.). Instead, rely fully on the AJAX search of the relationship field, and show ALL results that match the search as it's being typed.  (could probably place a max search result display to 99, as scrolling through 99 titles is as impractical as scrolling through thousands).  That way, there is no load/display/performance impact on the entry edit form from relationship fields, and the content editor then searches for all relationships.

Additionally, adjust the AJAX search so that it doesn't *start* search until the first character is entered in the search field.  Currently, it appears that if/when the display limit is set to 0 or 1, opening the relationship field takes a second (or maybe is just glitchy) because it appears that "nothing" entered into the search field is actually performing a search?  So, maybe require a keyup event after a character is entered into that search field before any AJAX searching is performed.

**Comment by shbchk** [comment]
It would be really helpful to have an option to show entry links and number of entries matching current search criteria:
![kjksss](https://user-images.githubusercontent.com/12234507/138360386-f29ca869-de8d-45fe-a78c-acb0f3f4a516.jpg)

**Comment by shbchk** [comment]
And an option to search everywhere, not just titles!

**Comment by blue-dreamer** [comment]
I was only thinking about this today!

**Ability to display multiple fields in the relationship field entries list**

I have a "people" channel where Title is used for the persons full name, "John Doe", I have a lot of entries and have three John Doe's, that means I can't choose the correct one. To solve this problem I need some other fields values so I can get the right match - "I'm looking for John Doe in [Canada] who's job is [Support technician]".

**Ability to use extra search filters**

I'm thinking generic data like category, status, it would be really nice to add filters for custom fields as well but that could be a head scratcher! That said maybe a popup advanced search type thing?

**Comment by robsonsobral** [comment]
I started to make little improvements here.

When using MSM, neither the channel filter nor the entries list show the site the item belong to.

The channel filter could have a kind of option label for the sites, grouping its channels.

Would a PR impact planned improvements? I guess not.

**Comment by intoeetive** [comment]
We are planning some UI changes for Relationship field (in particular, https://github.com/ExpressionEngine/ExpressionEngine/pull/1103 ) but I don't know whether that will be ready for 6.2 release.

As well we're adding min/max selection settings for the field (https://github.com/ExpressionEngine/ExpressionEngine/pull/1349) - which will not influence the UI of the field itself however

**Comment by shbchk** [comment]
That is good! 

Maybe, in addition to displaying the entry_id https://github.com/ExpressionEngine/ExpressionEngine/pull/1103, we can give it also a link to the entry page? I mean, we already have it in channel settings (Search results URL or Preview URL), why not use it in this case?

**Comment by pyrobob83** [comment]
Adding to this, can we please add a visual indication if already-related entries have since been closed like Playa used to do?  The simple green & red dots are very useful.

**Comment by intoeetive** [comment]
@pyrobob83 can you add a screenshot of what you're thinking about? (also separate new issue might work better for this  feature request)

**Comment by pyrobob83** [comment]
Done! https://github.com/ExpressionEngine/ExpressionEngine/issues/2522. But for what it's worth, I agree that the UX suggestions in this issue are worthwhile too, particularly links to edit the entries directly from the Publish screen where they're being related from.
